packages/@aws-cdk/aws-lambda/lib/edge-function.ts

```diff
@@ -0,0 +1,259 @@
+  public grantInvoke(identity: iam.IGrantable): iam.Grant {
+    return this.lambda.grantInvoke(identity);
+  }
+  public metric(metricName: string, props?: cloudwatch.MetricOptions): cloudwatch.Metric {
+    return this.lambda.metric(metricName, { ...props, region: EdgeFunction.EDGE_REGION });
+  }
+  public metricDuration(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
+    return this.lambda.metricDuration({ ...props, region: EdgeFunction.EDGE_REGION });
+  }
+  public metricErrors(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
+    return this.lambda.metricErrors({ ...props, region: EdgeFunction.EDGE_REGION });
+  }
+  public metricInvocations(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
+    return this.lambda.metricInvocations({ ...props, region: EdgeFunction.EDGE_REGION });
+  }
+  public metricThrottles(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
+    return this.lambda.metricThrottles({ ...props, region: EdgeFunction.EDGE_REGION });
+  }
+  public addEventSource(source: IEventSource): void {
+    return this.lambda.addEventSource(source);
+  }
+  public configureAsyncInvoke(options: EventInvokeConfigOptions): void {
+    return this.lambda.configureAsyncInvoke(options);
+  }
+
+  /** Create a function in-region */
+  private createInRegionFunction(id: string, props: FunctionProps): FunctionConfig {
+    const role = props.role ?? defaultLambdaRole(this, id);
+    const edgeFunction = new Function(this, 'Fn', {
+      ...props,
+      role,
+    });
+
+    return { edgeFunction, edgeArn: edgeFunction.currentVersion.edgeArn, functionStack: this.stack };
+  }
+
+  /** Create a support stack and function in us-east-1, and a SSM reader in-region */
+  private createCrossRegionFunction(id: string, props: FunctionProps): FunctionConfig {
+    const parameterName = `EdgeFunctionArn${id}`;
+    const functionStack = this.edgeStack();
+    this.stack.addDependency(functionStack);
+
+    const edgeFunction = functionStack.addEdgeFunction(id, parameterName, props);
+
+    const parameterArn = this.stack.formatArn({
+      service: 'ssm',
+      region: EdgeFunction.EDGE_REGION,
+      resource: 'parameter',
+      resourceName: parameterName,
+    });
+
+    const resourceType = 'Custom::CrossRegionStringParameterReader';
+    const serviceToken = CustomResourceProvider.getOrCreate(this, resourceType, {
+      codeDirectory: path.join(__dirname, 'edge-function'),
+      runtime: CustomResourceProviderRuntime.NODEJS_12,
+      policyStatements: [{
+        Effect: 'Allow',
+        Resource: parameterArn,
+        Action: ['ssm:GetParameter'],
+      }],
+    });
+    const resource = new CustomResource(this, 'ArnReader', {
+      resourceType: resourceType,
+      serviceToken,
+      properties: {
+        Region: EdgeFunction.EDGE_REGION,
+        ParameterName: parameterName,
+        RefreshEachDeploy: Date.now().toString(), // Ensure this value is refreshed on each deploy, to get the latest function ARN.
```

Comment:
This is going to cause the CloudFormation template to modify on each synth. We don't do this! One of the core operating principles of the CDK is that updates must be triggered by a source change.

Instead, we need to find a way to trigger the downstream CloudFormation deployment if the parameter value changes. How about using the hash of the function ARN in the parameter name?

Reply:
I'm glad you pointed this out; it's one of the main points I want feedback on here.

The problem here is that the whole 'trick' to this working is defining both a parameter name and update trigger that aren't reliant on the `Function` itself. As soon as the name/token rely on the `Function` (e.g., function ARN), we run into the very cross-region reference issue that prompts the creation of this construct in the first place.

Ideally, this token -- rather than a `Date` -- would be the consistent source/asset hash of the Function's code. However, I couldn't find any way to get at that value in (a) a consistent way and (b) without requiring the Function to be created/referenced first.

We could add a(n optional?) property to `Code` called `codeId` or `codeHash` or something, that returned a (pseudo-)hash of the code (whether S3/Params/Asset/Inline) that could be used to trigger an update. This looks like it would require some non-trivial refactoring though and I wasn't willing to dive deeply into that without an up-front design discussion.

Reply:
_(Capturing offline conversation)_

Found a good solution in the `calculateFunctionHash` function, which doesn't appear to cause cross-region issues.